amf-api-contract/shared/src/main/scala/amf/apicontract/internal/transformation/stages/GraphQLInterfaceRecursionDetectionStage.scala

```diff
@@ -0,0 +1,67 @@
+package amf.apicontract.internal.transformation.stages
+
+import amf.apicontract.internal.validation.definitions.ResolutionSideValidations.RecursiveInheritance
+import amf.core.client.scala.AMFGraphConfiguration
+import amf.core.client.scala.errorhandling.AMFErrorHandler
+import amf.core.client.scala.model.document.{BaseUnit, DeclaresModel}
+import amf.core.client.scala.model.domain.{AmfArray, RecursiveShape}
+import amf.core.client.scala.transform.TransformationStep
+import amf.core.internal.metamodel.domain.ShapeModel
+import amf.core.internal.metamodel.domain.ShapeModel.Inherits
+import amf.core.internal.parser.domain.Value
+import amf.core.internal.unsafe.PlatformSecrets
+import amf.shapes.client.scala.model.domain.NodeShape
+
+case class GraphQLInterfaceRecursionDetectionStage() extends TransformationStep() with PlatformSecrets {
+
+  override def transform(
+      model: BaseUnit,
+      errorHandler: AMFErrorHandler,
+      configuration: AMFGraphConfiguration
+  ): BaseUnit = {
+    implicit val eh: AMFErrorHandler = errorHandler
+    model match {
+      case d: DeclaresModel =>
+        d.declares.foreach {
+          case node: NodeShape if isInterface(node) => traverse(node) // only interfaces might have cyclic inheritance
+          case _                                    =>                // ignore
+        }
+      case _ => // ignore
+    }
+    model
+  }
+
+  private def isInterface(node: NodeShape): Boolean = node.isAbstract.value()
+
+  def traverse(current: NodeShape, previous: Seq[NodeShape] = Nil)(implicit eh: AMFErrorHandler): Unit = {
+    // Place recursive shapes in current level
+    val newInherits = current.inherits.map {
+      case next: NodeShape if isInterface(next) && previous.contains(next) =>
+        reportError(current, previous, next, eh)
+        RecursiveShape(next)
```

Comment:
Since you should set the inherits field without re-adopting, you should adopt the newly created Recursive shape here